### ISSUE TITLE
Database Deadlock on YTAB_EXEMPTIONS

### ISSUE THREAD

**Issue by abap-weasel** [SAP/code-pal-for-abap#159]
Hi Code Pal Team,

we scheduled Code Pal check for large custom code base as ATC run using multiple processes (see bottom right in screenshot).
After several hours we noticed deadlock on database YTAB_EXEMPTIONS (see top, bottom left) and we had to cancel ATC run.

Kindly check and advise.

![grafik](https://user-images.githubusercontent.com/47646214/94725162-a03a1380-035b-11eb-951a-3bfb40bc4a39.png)

**Comment by lucasborin** [comment]
Could you please replace the `insert_created_on_to_buffer` method from the `y_object_creation_date` class with the following code and test it again?  

```abap
  METHOD insert_created_on_to_buffer.
    DATA: exemption TYPE ytab_exemptions.

    exemption-object = object_type.
    exemption-obj_name = object_name.
    exemption-created_on = creation_date.
    exemption-as4date_co = sy-datum.
    exemption-is_created_on_buffered = abap_true.

    MODIFY ytab_exemptions FROM exemption.

    is_stored = xsdbool( sy-subrc = 0 ).
  ENDMETHOD.
```

**Comment by abap-weasel** [comment]
Unfortunately issue is still happening with MODIFY statement. Also check for record still leads to deadlock.
I would expect that a COMMIT WORK after DB operation removes lock. Is it safe to try that or is that not allowed in ATC context?

![grafik](https://user-images.githubusercontent.com/47646214/94862742-b0282500-0439-11eb-8e73-bd7f6f780541.png)

**Comment by lucasborin** [comment]
I think it is 'safe' if you are executing the pal code only. However, I am not sure if it will solve the issue. You can try it in your development landscape.

To solve this issue, I can provide you a quick workaround: Disabling the buffer. 

You just have to replace the `y_object_creation_date` with the following one:
[y_object_creation_date.txt](https://github.com/SAP/code-pal-for-abap/files/5319177/y_object_creation_date.txt)

Furthermore, it looks like this issue will demand a little time/effort since we have to refactor the buffer, and, maybe, split the `ytab_exemption` responsibility in two as it buffers the object creation date and stores exempted object names. 

Ps: Feel free to contribute to the project sharing ideas on how to refactor the buffer routine. :)

**Comment by abap-weasel** [comment]
I am currently testing your version of y_object_creation_date with deactivated buffer.
Wondering about another deadlock on UPDATE DDNTT. Any idea what is happening there?

![grafik](https://user-images.githubusercontent.com/47646214/95117386-d0612800-0748-11eb-8684-c394134c2858.png)

As only two tasks are blocked by that I expect to have the check run through after some time. 
Will keep you updated on that, but wanted to let you now the new observation for now.

**Comment by abap-weasel** [comment]
Unfortunately more and more tasks got blocked again, so I had to abort the ATC run again.

While doing so I did some further investiation. 
Before I cancelled the task that had HOLD status, I started process debugging in SM50 for first WAIT task.
After I cancelled the HOLD task, this is there debugger stopped:

![grafik](https://user-images.githubusercontent.com/47646214/95118832-3b136300-074b-11eb-9356-fbc3f77f61c9.png)

Now I am even more puzzled as I expected to find some workbench API but not the INSERT on the ytab_exemptions as that strange DDNTT table was shown in lock monitor. Maybe you can provide some insights on this. 😕

**Comment by abap-weasel** [comment]
Well, I have some news on this. I have worked around the issue by schedulung ATC run with a single process. Job completed after about 15 days (yes, days! 😁).

**Comment by lucasborin** [comment]
I'll try to implement a static buffer (not DB based) using shared memory. I will let you know when the code is ready for the beta test.

**Comment by abap-weasel** [comment]
It took some time, but I can confirm that new version from database-deadlock branch is working. ATC job ran for about 46 hours with 15 tasks.

**Comment by lucasborin** [comment]
It is a good start. Now, we can work on performance improvement.

**Comment by lucasborin** [comment]
Solution merged into next version, under branch 1.07.0

**Comment by lucasborin** [comment]
@abap-weasel: I kindly ask you to open a new incident with a performance trace (high-level is enough) so I can have a look and work on the findings. Besides, I ran it locally and most of the time was in the code inspector framework side and not in the code pal. 

Would you mind if I close this incident?

**Comment by abap-weasel** [comment]
Yeah, fine with me (as change was just merged into master).
